clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -10570,6 +10570,26 @@ Value *CodeGenFunction::EmitAArch64BuiltinExpr(unsigned BuiltinID,
     return Builder.CreateCall(F, llvm::ConstantInt::get(Int32Ty, HintID));
   }
 
+  if (BuiltinID == clang::AArch64::BI__builtin_arm_get_sme_state) {
+    // Create call to __arm_sme_state and store the results to the two pointers.
```

Comment:
Out of interest why does this builtin return the result via memory? instead of being an alias of `__arm_sme_state`. Or rather, can `__arm_has_sme` and `__arm_in_streaming_mode` call `__arm_sme_state` directly?

Reply:
It could call `__arm_sme_state` directly, but we don't expose the special calling convention for __arm_sme_state with an attribute at the C level, hence the indirection through a builtin.

Reply:
OK, that makes sense.  But the new builtin can still return by value?

Reply:
I looked into this a while back when I implemented this builtin, but couldn't find any prior work where a builtin returned two scalar values which is why I opted for doing this through memory. LLVM will optimise these away similar to how it optimises all of Clang's other memory operations away (mem2reg).

Reply:
I see what you mean. In which case I thinking creating specific builtins to avoid having to go through memory would be better because I don’t get the point of creating a builtin that has a less efficient interface than the function it targets.  That said, I’ll not push for it if you feel this implementation is better.